### PR TITLE
INT-3001 - Ingest hosts

### PR DESCRIPTION
**Pull request by eXtremeX**
- Updated `iterateApi` to handle both V1 and V2 of Datadog API Client
- The following entities are **now** created:

| Resources | Entity `_type` | Entity `_class` |
| --------- | -------------- | --------------- |
| Host      | `datadog_host` | `Host`          |

- The following relationships are **now** created:

| Source Entity `_type` | Relationship `_class` | Target Entity `_type` |
| --------------------- | --------------------- | --------------------- |
| `datadog_account`     | **HAS**               | `datadog_host`        |

### CODE REVIEW

File: src/steps/hosts/__recordings__/fetchHosts_2528439642/recording.har [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+              "name": "x-content-type-options",
+              "value": "nosniff"
+            },
+            {
+              "name": "strict-transport-security",
+              "value": "max-age=15724800;"
+            },
+            {
+              "name": "content-security-policy",
+              "value": "frame-ancestors 'self'; report-uri https://logs.browser-intake-datadoghq.com/api/v2/logs?dd-api-key=pube4f163c23bbf91c16b8f57f56af9fc58&dd-evp-origin=content-security-policy&ddsource=csp-report&ddtags=site%3Adatadoghq.com"
```
Comment (dobregons) [inline src/steps/hosts/__recordings__/fetchHosts_2528439642/recording.har:94]: Shouldn't this be redacted? This is exposing the value of `dd-api-key`

File: src/client.ts [hunk 2]
```diff
@@ -61,14 +64,15 @@ export class APIClient {
...
-          await callback(res.data);
+        if (res[dataParam]) {
+          await callback(res[dataParam]);
         }
 
         totalCount = res.meta?.page?.totalCount || 0;
         processed += this.pageSize;
         pageNumber += 1;
       } catch (err) {
+        console.error(err);
```
Comment (zemberdotnet) [inline src/client.ts:75]:
```suggestion
```

File: src/client.ts [hunk 3]
```diff
@@ -30,6 +32,7 @@ export class APIClient {
   private async iterateApi<T, K>(
     client: K,
     fnName: string,
+    dataParam: string,
```
Comment (zemberdotnet) [inline src/client.ts:35]: It looks like this is good, but I'm a little confused what the purpose of `dataParam` is. Is it to select data from the response?

In that case a name like `responseDataProperty` may be clearer.

File: src/steps/hosts/converters.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        'meta.agentVersion': host.meta?.agentVersion,
+        'meta.cpuCores': host.meta?.cpuCores,
+        'meta.machine': host.meta?.machine,
+        'meta.nixV': host.meta?.nixV,
+        'meta.platform': host.meta?.platform,
+        'meta.processor': host.meta?.processor,
+        'meta.pythonV': host.meta?.pythonV,
+        'meta.socketFqdn': host.meta?.socketFqdn,
+        'meta.socketHostname': host.meta?.socketHostname,
+        'meta.winV': host.meta?.winV,
```
Comment (zemberdotnet) [inline src/steps/hosts/converters.ts:41]: I think we should give these camelCase names. I've made suggestions.
```suggestion
        agentVersion: host.meta?.agentVersion,
        hostCpuCores: host.meta?.cpuCores,
        machine: host.meta?.machine,
        platform: host.meta?.platform,
        processor: host.meta?.processor,
        pythonVersion: host.meta?.pythonV,
        socketFqdn: host.meta?.socketFqdn,
        socketHostname: host.meta?.socketHostname,
```

File: src/steps/hosts/converters.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    entityData: {
+      source: host,
+      assign: {
+        _type: Entities.HOST._type,
+        _class: Entities.HOST._class,
+        _key: getHostKey(host.id!),
+        id: `${host.id}`,
+        aliases: host.aliases,
+        apps: host.apps,
+        awsName: host.awsName,
```
Comment (zemberdotnet) [inline src/steps/hosts/converters.ts:25]: What is `awsName`?

File: test/mocks.ts [hunk 6]
```diff
@@ -94,3 +97,40 @@ export function getMockUser(): Partial<User_v2> {
...
+      macV: ['', '', ''],
+      machine: 'amd64',
+      nixV: ['ubuntu', '22.04', ''],
+      platform: 'linux',
+      processor: 'AMD FX(tm)-8350 Eight-Core Processor',
+      pythonV: '3.8.14',
+      socketFqdn: '82c729d4227d',
+      socketHostname: '82c729d4227d',
+      winV: ['', '', ''],
+    },
```
Comment (zemberdotnet) [inline test/mocks.ts:129]: We should double check that this is actually the structure of the response. I went through the response and the response looked different.

File: src/steps/hosts/converters.ts [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+  parseTimePropertyValue,
+} from '@jupiterone/integration-sdk-core';
+
+import { Entities } from '../constants';
+
+export function getHostKey(id: number): string {
+  return `datadog_host:${id}`;
+}
+
+export function createHostEntity(host: Host): Entity | null {
```
Comment (zemberdotnet) [inline src/steps/hosts/converters.ts:14]: Why does this return `Entity | null`? It looks like it only ever returns an `Entity`.

File: src/steps/hosts/converters.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        hostname: host.hostName,
+        isMuted: host.isMuted,
+        reportedOn: parseTimePropertyValue(host.lastReportedTime),
+        name: host.name,
+        sources: host.sources,
+        state: host.up ? 'running' : 'stopped',
+        'meta.agentVersion': host.meta?.agentVersion,
+        'meta.cpuCores': host.meta?.cpuCores,
+        'meta.machine': host.meta?.machine,
+        'meta.nixV': host.meta?.nixV,
```
Comment (zemberdotnet) [inline src/steps/hosts/converters.ts:35]: nixV, winV, and macV should all be used to fill in properties like `osName` and `osVersion`.
E.g.
If `nixV` is:"nixV":["ubuntu","22.04",""], then we need to make properties `osName: "ubuntu"`, `osVersion: "22.04"`.

We should do the same for macV and winV.

File: src/steps/hosts/converters.ts (hunk 4 above)
Comment (zemberdotnet) [reply src/steps/hosts/converters.ts:41]: I've removed `winV` and `nixV` in reference to my comment above. We should also double check that this is the right structure. The raw response looks like some of these properties my be more deeply nested.

File: test/mocks.ts (hunk 6 above)
Comment (eXtremeX) [reply test/mocks.ts]: The recording's content does look different - e.g. if you were to search for `socketFqdn` in the recording.har you wouldn't find it, however if you run the test (offline) against the stored recording and output the found host(s), it does show fields exactly as the converter expects them.

**Review by zemberdotnet — APPROVED** [review]

**Comment by zemberdotnet** [comment]
@eXtremeX good to go needs a release

**Pull request merged**